testCPUID-all/testCPUID.cpp

```diff
@@ -0,0 +1,433 @@
+
+    while (g_aligned_global_location.loc == original_value)
+    {
+        //_umonitor(&g_aligned_global_location.loc);
+
+        //_umwait(1, g_timeout);
+        //_tpause(0, g_timeout);
+        _tpause(1, g_timeout);
+        //printf("wait exited\n");
+        waited_count++;
+    }
+
+    //printf("[%10s] changed to %Id and waited %s times!\n", "tpause", g_aligned_global_location.loc, formatNumber(waited_count));
+    g_waited_count = waited_count;
+    return 0;
+}
+
+DWORD WINAPI ThreadFunction_umwait(LPVOID lpParam)
+{
+    size_t original_value = g_aligned_global_location.loc;
+    size_t waited_count = 0;
+    printf("original value is %Id, timeout is %I64d\n", original_value, g_timeout);
+
+    while (g_aligned_global_location.loc == original_value)
+    {
+        _umonitor((void*)&g_aligned_global_location.loc);
+        _umwait(1, g_timeout);
+        //printf("wait exited\n");
+        waited_count++;
+    }
+
+    //printf("[%10s] changed to %Id and waited %s times!\n", "umwait", g_aligned_global_location.loc, formatNumber(waited_count));
+
+    g_waited_count = waited_count;
+    return 0;
+}
+
+DWORD WINAPI ThreadFunction_mwaitx(LPVOID lpParam)
+{
+    size_t original_value = g_aligned_global_location.loc;
+    size_t waited_count = 0;
+    printf("original value is %Id, timeout is %I64d\n", original_value, g_timeout);
+    uint64_t start_cycles = __rdtsc();
+    auto time_start = std::chrono::steady_clock::now();
+
+    while (g_aligned_global_location.loc == original_value)
+    {
+        _mm_monitorx((const void*)&g_aligned_global_location.loc, 0, 0);
+
+        if (g_aligned_global_location.loc == original_value) {
+            waited_count++;
+            _mm_mwaitx(2, 0, (uint32_t)g_timeout);
+        }
+    }
+
+    g_total_cycles_on_thread = __rdtsc() - start_cycles;
+    g_elapsed_time_us_on_thread = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now() - time_start).count();
+
+    printf("[%10s] changed to %Id and waited %s times!\n", "mwaitx", g_aligned_global_location.loc, formatNumber(waited_count));
+
+    g_waited_count = waited_count;
+    return 0;
+}
+
+DWORD WINAPI ThreadFunction_mwaitx2(LPVOID lpParam)
+{
+    size_t original_value = g_aligned_global_location.loc;
+    size_t waited_count = 0;
+    printf("original value is %Id, timeout is %I64d\n", original_value, g_timeout);
+
+    while (g_aligned_global_location.loc == original_value)
+    {
+        _mm_monitorx((const void*)&g_aligned_global_location.loc, 0, 0);
+
+        if (g_aligned_global_location.loc == original_value) {
+            _mm_mwaitx(2, 0, (uint32_t)g_timeout);
+            waited_count++;
+        }
+    }
+
+    printf("[%10s] changed to %Id and waited %s times!\n", "mwaitx", g_aligned_global_location.loc, formatNumber(waited_count));
+
+    g_waited_count = waited_count;
+    return 0;
+}
+
+uint64_t g_total_iter = 10000;
+uint64_t sum = 0;
+
+void inc_only()
```

Comment:
Do you want to keep `inc_with_pause()` and `inc_only()`?

Reply:
i would keep it..it's good data to have.